runtime/lua/vim/lsp/client.lua

```diff
@@ -0,0 +1,663 @@
+---
+--- Sends a notification to an LSP server.
+--- Returns: a boolean to indicate if the notification was successful. If
+--- it is false, then it will always be false (the client has shutdown).
+--- @field notify fun(method: string, params: table?): boolean
+---
+--- Cancels a request with a given request id.
+--- Returns: same as `notify()`.
+--- @field cancel_request fun(id: integer): boolean
+---
+--- Stops a client, optionally with force.
+--- By default, it will just ask the server to shutdown without force.
+--- If you request to stop a client which has previously been requested to
+--- shutdown, it will automatically escalate and force shutdown.
+--- @field stop fun(force?: boolean)
+---
+--- Runs the on_attach function from the client's config if it was defined.
+--- Useful for buffer-local setup.
+--- @field on_attach fun(bufnr: integer)
+---
+--- Checks if a client supports a given method.
+--- Always returns true for unknown off-spec methods.
+--- [opts] is a optional `{bufnr?: integer}` table.
+--- Some language server capabilities can be file specific.
+--- @field supports_method fun(method: string, opts?: {bufnr: integer?}): boolean
+---
+--- Checks whether a client is stopped.
+--- Returns: true if the client is fully stopped.
+--- @field is_stopped fun(): boolean
+local Client = {}
+Client.__index = Client
+
+--- @param cls table
+--- @param meth any
+--- @return function
+local function method_wrapper(cls, meth)
+  return function(...)
+    return meth(cls, ...)
+  end
+end
+
+--- @package
+--- @param id integer
+--- @param rpc vim.lsp.rpc.PublicClient
+--- @param handlers table<string,lsp.Handler>
+--- @param offset_encoding string
+--- @param config lsp.ClientConfig
+--- @return lsp.Client
+function Client.new(id, rpc, handlers, offset_encoding, config)
+  local name = config.name
+
+  --- @class lsp.Client
+  local self = {
+    id = id,
+    config = config,
+    handlers = handlers,
+    rpc = rpc,
+    offset_encoding = offset_encoding,
+    name = name,
+    _log_prefix = string.format('LSP[%s]', name),
+    requests = {},
+    commands = config.commands or {},
+    attached_buffers = {},
+    server_capabilities = {},
+    dynamic_capabilities = vim.lsp._dynamic.new(id),
+
+    --- Contains $/progress report messages.
+    --- They have the format {token: integer|string, value: any}
+    --- For "work done progress", value will be one of:
+    --- - lsp.WorkDoneProgressBegin,
+    --- - lsp.WorkDoneProgressReport (extended with title from Begin)
+    --- - lsp.WorkDoneProgressEnd    (extended with title from Begin)
+    progress = vim.ringbuf(50) --[[@as lsp.Client.Progress]],
+
+    --- @deprecated use client.progress instead
+    messages = { name = name, messages = {}, progress = {}, status = {} },
+  }
+
+  self.request = method_wrapper(self, Client._request)
```

Comment:
Nice, bwc concerns were the main reason I had never attempted this. Nice solution